### ISSUE TITLE
Inconsistent types between optional chaining and explicit if statement

### ISSUE THREAD

**Issue by danvk** [microsoft/TypeScript#36163]
**TypeScript Version:**  Nightly

**Search Terms:** strictNullChecks uninitialized

**Expected behavior:** The two samples below should either both produce an error (my preference) or neither should produce an error.

**Actual behavior:** There's an error for the explicit if statement but not the optional chain.

**Related Issues:** #30053

**Code**
```ts
class MyClass {
  getFeatures() {
    return [1, 2, 3];
  }
}

let mc: MyClass;

function query() {
  if (mc !== null && mc !== undefined) {
    return mc.getFeatures();
  } else {
    return undefined;
  }
}

console.log(query().length);
         // ~~~~~~~ Error: Object is possibly 'undefined'. (2532)
```

on the other hand:

```ts
class MyClass {
  getFeatures() {
    return [1, 2, 3];
  }
}

let mc: MyClass;

function query() {
  return mc?.getFeatures();
}

console.log(query().length);  // no error, throws at runtime
```

I would expect these to both be errors with `strictNullChecks` enabled. I ran into a `TypeError` in production today that I traced back to this!

<details><summary><b>Output</b></summary>

```ts
"use strict";
class MyClass {
    getFeatures() {
        return [1, 2, 3];
    }
}
let mc;
function query() {
    if (mc !== null && mc !== undefined) {
        return mc.getFeatures();
    }
    else {
        return undefined;
    }
}
console.log(query().length);
```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/?ts=3.8.0-dev.20200111#code/MYGwhgzhAECyCeBhcVoG8CwAoa0DmApgC4BiBYRArgE4EQAUAlOtrrrVdQHbQDaAjABpoAJmEBmALoBuVtAC+2RVmwhi0ALbAAXHCQoIslVgBmlLsCIBLAPY8AjpQLV4TFjmhWT0elugBCAF5A6C5KEBBoADIozWAA4OhzABMCEysuAmTmTA92YhoeLQA6QlJyTjomI1x5aAIQCAJ3NmgOQqSuVPTM5JqFJWxsYDsIGzVikBs8ekdnV0ZJgi48IgALRiMgA)

**Comment by RyanCavanaugh** [comment]
There's an error in the nightly playground, as expected - maybe this just got fixed?

**Comment by danvk** [comment]
@RyanCavanaugh I don't think that's it, I see the same issue with 3.7.2. Here's another [playground link][1].

[1]: https://www.typescriptlang.org/play/?ssl=23&ssc=36&pln=23&pc=44#code/MYGwhgzhAECyCeBhcVoG8CwAoa0DmApgC4BiBYRArgE4EQAUAlOtrrrVdQHbQDaAjABpoAJmEBmALoBuVtAC+2RVmwhi0ALbAAXHCQoIslVgBmlLsCIBLAPY8AjpQLUkACzBWuTFjmgcaPFoA-AB0hKTknHRMRsrYwHYQNmohIDZ49I7Obh5ejKkEXHhErozSuAD0FdBcNtDO1DbU2NhmFtZ20FkuAJIm3pi+VibQ9FrQALxTNZQgINAAPguawJPT5gAmBCaeBBvMg2x+xAHQm9u7G0a48vUgEAQ+R-7cK2HEZBQ00WVyynFYBJcJIpNIZbrwPpMApFEplSrVADyACMAFYESzQKwwAAONigVmRIHg0AA5Ocdlw9qSQtggA

**Comment by RyanCavanaugh** [comment]
@danvk that link shows only expected behavior. Inside a closure, `let`s are assumed to have been already initialized.

**Comment by danvk** [comment]
@RyanCavanaugh I see the same behavior for both nightly and 3.7.2, i.e. an error for the version with an `if` statement but not the version with a conditional chain:

![image](https://user-images.githubusercontent.com/98301/72357057-52d82200-36b8-11ea-8e12-e72c02189c55.png)

Here's a version which shows different behavior for an optional chain and an equivalent ternary, which I would really expect to be the same! [Playground link][1]

```ts
// ...
let mc: MyClass;

function queryChain() {
  return mc?.getFeatures();
}

function queryConditional() {
  return (mc === null || mc === undefined) ? undefined : mc.getFeatures();
}

console.log(queryChain().length);  // no error
console.log(queryConditional().length);  // Object is possibly 'undefined'.

```

[1]: https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=19&pc=1#code/MYGwhgzhAECyCeBhcVoG8CwAoa0DmApgC4BiBYRArgE4EQAUAlOtrrrVdQHbQDaAjABpoAJmEBmALoBuVtAC+2RVmwhi0ALbAAXHCQoIslVgBmlLsCIBLAPY8AjpQLUkACzBWuTFjmgcaPFoA-AB0hKTknHRMRsrYZhbWdtCOzkh2ACZWSVxgIN6Yvv7c0PRa0AC8VdBclCAg0AA+jZrAldXmGQQmngQZzEHQnd29GdC6WmHEZBQ00Yyx2NjAdhA2aiEgNnj0qS6I7p5MmwRceESuC7gA9Nc1NtDO1DbUy6vrBJvbu077mdm2XL5RgnM4XK7QW7QADyACMAFYESzQKwwAAONigVlhIHg0AA5MMelw+viQtggA

**Comment by RyanCavanaugh** [comment]
So I'm not sure which one of these should be considered to be "wrong", but these are the relevant designed behaviors, all of which are working as intended:

 * Defensive comparisons to `null` and `undefined` are always allowed
 * Both branches of a conditional are assumed to be reachable
 * When the CFA graph doesn't prove otherwise, `let`s are assumed to have been "already" initialized to their annotated type
 * `e?.x` does not introduce `| undefined` to the type of `e.x` when we believe `e` is never `null` / `undefined`
 * `e?.x` is not an error when we believe `e` is never `null` / `undefined`

**Comment by danvk** [comment]
This is the one that surprises me:

- e?.x does not introduce | undefined to the type of e.x when we believe e is never null / undefined

That makes sense in isolation, but is the opposite behavior from what you get in a ternary which an optional chain is just shorthand for. Is there a good rationale for divergent behaviors in these two cases?

```ts
let a = mc?.getFeatures();  // number[]
let b = (mc === null || mc === undefined) ? undefined : mc.getFeatures();  // number[] | undefined
```

**Comment by RyanCavanaugh** [comment]
Don't tell me they should be the same; tell me which one you think should change 😉

**Comment by danvk** [comment]
I get to choose? 🙀

With `strictNullChecks` I'd prefer that `mc?.getFeatures()` include the `| undefined`. Presumably there's some reason you've chosen to include the optional chain. If you want to keep it at runtime but don't want it to affect the types, then you should be forced to add a type assertion.

**Comment by RyanCavanaugh** [comment]
> If you want to keep it at runtime but don't want it to affect the types, then you should be forced to add a type assertion.

I was thinking about this a bunch today and am having a hard time convincing myself that this isn't correct. It's unfortunate we shipped it the other way already, but maybe it'd be an acceptable break.

**Comment by danvk** [comment]
@RyanCavanaugh did you want this to be closed? Has a decision been made one way or the other?
